allennlp/nn/beam_search.py

```diff
@@ -0,0 +1,262 @@
+            should be a tensor of shape ``(batch_size, *)``, where ``*`` means any other
+            number of dimensions.
+        step : ``StepFunctionType``
+            A function that is responsible for computing the next most likely tokens,
+            given the current state and the predictions from the last time step.
+            The function should accept two arguments. The first being a tensor
+            of shape ``(group_size,)``, representing the index of the predicted
+            tokens from the last time step, and the second being the current state.
+            The ``group_size`` will be ``batch_size * beam_size``, except in the initial
+            step, for which it will just be ``batch_size``.
+            The function is expected to return a tuple, where the first element
+            is a tensor of shape ``(group_size, target_vocab_size)`` containing
+            the log probabilities of the tokens for the next step, and the second
+            element is the updated state. The tensor in the state should have shape
+            ``(group_size, *)``, where ``*`` means any other number of dimensions.
+        first_step : ``StepFunctionType``, optional
+            If the first step of decoding should be handled differently, then you can
+            set this function which will only be used during the first step. If not set,
+            ``step`` will be used for the first step as well. This function should have the
+            same signature as ``step``.
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            Tuple of ``(predictions, log_probabilities)``, where ``predictions``
+            has shape ``(batch_size, beam_size, max_steps)`` and ``log_probabilities``
+            has shape ``(batch_size, beam_size)``.
+        """
+        batch_size = start_predictions.size()[0]
+
+        first_step = first_step or step
+
+        # List of (batch_size, beam_size) tensors. One for each time step. Does not
+        # include the start symbols, which are implicit.
+        predictions: List[torch.Tensor] = []
+
+        # List of (batch_size, beam_size) tensors. One for each time step. None for
+        # the first.  Stores the index n for the parent prediction, i.e.
+        # predictions[t-1][i][n], that it came from.
+        backpointers: List[torch.Tensor] = []
+
+        # Calculate the first timestep. This is done outside the main loop
+        # because we are going from a single decoder input (the output from the
+        # encoder) to the top `beam_size` decoder outputs. On the other hand,
+        # within the main loop we are going from the `beam_size` elements of the
+        # beam to `beam_size`^2 candidates from which we will select the top
+        # `beam_size` elements for the next iteration.
+        start_class_log_probabilities, state = first_step(start_predictions, start_state)
```

Comment:
It doesn't seem like an explicit `first_step` is being passed in anymore, so I'm a bit confused as to how the event2mind model would continue to work. Could you clarify? Sorry if I'm missing something obvious here!

Reply:
Sorry, got confused here. This is definitely working. :) Is there actually a use case for having `first_step` separate from `step` presently?